### DISCUSSION TITLE
CANbus networking support

### DISCUSSION THREAD

**Discussion by NSD76** [rsta2/circle#473] [Ideas]
Hello,
I would like to use Circle as basis system for project on RPI  3B and later boards  needing CAN bus networking (using usual RPI MCP2515 -based HAT). Is there anything already started by someone for such feature ?

Otherwise I may start to write some driver port for this system. I'm not used to program for Circle (this is my first project with this library), but used to develop low-level drivers for almost any kind of usual embedded peripherals based on devices datasheets/users manuals.

**Comment by rsta2** [comment]
I'm not aware of any CAN bus support with Circle, but it would be great to have it. Let me know, when you need support for this.

**Reply by NSD76** [reply]
Thanks for feedback, then I will start to develop something. 

I wanted initially to make a port of existing CAN WiringPI library but I find it too complicated.  I feel more comfortable starting from almost scratch, at least from some sample code (or at least register definitions) proposed by Microchip for use on PIC MCUs, but  mostly by using MCP2515 datasheet infos.

I may post soon some functional description of the library for each of the exposed functions of the Lib I propose (as pseudo code for the moment) before going into the details, for the moment I will put this in my own repository on git (I will provide the link in an upcoming post), but once the sate would be more advanced I may propose to submit this in Circle external contributors section.

I might need some support once I will have to make properly the link with Circle libraries (GPIO, SPI...), but I will trw first on my side.
Of course doing this as hobby, so it will for sure take some time ;-)

**Reply by dlbogdan** [reply]
Sounds great! I wish I had time to help (2 smol kids). For now at least you have me as a supporter.

**Reply by rsta2** [reply]
Sounds like a good plan. Take the time, you need. Let me know, when you have questions.

**Reply by NSD76** [reply]
Thank you for encouraging !
I  did already start describing some principles for the lib and how I would like to implement it. I'm not used to play with Git (this is my first project with github), so maybe at some point I will raise questions about how to share, post...   I already posted 2 description files in my own repo but don't have found yet how I could give access to. Maybe today I will take more time to search.

Now currently evaluating if I prefer re write  from scratch an HAL for the CAN device MCP2515, or if  I try to port an existing one made for another board or system (e.g. there are libs for Arduino boards available on Github). 
I would like the HAL be if possible not dependant from host machine (so not specific to the RPI and to Circle), and then put in between the necessary wrapper functions / objects for the use from a Circle app.

->The question of porting existing lib and integrate the ported code into Circle also raises question of possible License mismatch between ported lib licence (usually MIT) and Circle one (GNU). Could somebody advise on that point ? 
As info for me it makes anyway no difference in amount of work between porting existing lib (so taking time to understand what has been coded then adapting)  or starting new from scratch since here the device is not so complicated.

**Reply by rsta2** [reply]
You are welcome. I'm not a lawyer, but [Wikipedia](https://en.wikipedia.org/wiki/MIT_License) tells (in the summary box), the MIT License is compatible with GPL.

**Reply by NSD76** [reply]
I have created a repo under my name for the moment (NSD76/CANlib-MCP2515-for-Circle), still private until I get true C/C++ code  to submit. 
As temporary I invite all who replied in this discussion topic (you should have got an invitation).
For the moment I mainly work on the overall principle and organization of the lib. If you can have a quick look at the files your comments are welcome either here or under the discussion list on my repository, already committed 3 drafts of documents I started to write yesterday (clarity and presentation may not be top, this are mainly first ideas I roughly put on these docs). I assume I will go on that way for some next weeks by refining and structuring this, before really starting to code.

I have a friend that is interested to help me (already as collaborator  on my repository), for the background project for which I need CAN cupport with Circle (will speal about this later, another repo will be created for it once ready to start), so maybe he will also support for this part.

Regarding license scheme I may use the MIT at least for the HAL since I read that is it less constraining than the GPL while being compatible with (portion of code can be inluded into another code using GPL, but not the opposite). Could ease the re-use by somebody else if needed. For the Circle Wrapper code (using the HAL) then GPL makes more sense since it can be only used with/for Circle based app.

**Reply by rsta2** [reply]
I will have a look into your files soon.

**Reply by rsta2** [reply]
I had a look into your files. I think, that's OK so far. I cannot speak much about CAN bus, because I'm not familiar with it, but from Circle point of view, you should have a look at the following classes, which are abstractions for the hardware, which must be accessed for the CAN bus HAT, as I understand it:

* `CSPIMasterDMA` (DMA-driven SPI access, polling for short packets supported)
* `CGPIOPin` (/INT input from MCP2515)
* `CGPIOManager` (necessary to generate interrupts from `CGPIOPin`)
* maybe `CTimer` (for timeouts etc.)

These classes are described [here](https://circle-rpi.readthedocs.io/en/47.0/basic-system-services/gpio-access.html). IRQ-driven SPI access is not supported by Circle, but not necessary, because with DMA it can run completely in the background for larger transfers.

Please note, that throughout Circle GPIO numbers are SoC numbers (e.g. /INT is connected to GPIO25 here, while its pin 22 on the GPIO header).

BTW. The Discussions in your repo are not enabled yet.

**Reply by NSD76** [reply]
Hello,
thanks for feedback. I forgot the point about DMA support in Circle, agree that it should definitely help if multiple successive SPI transfers can use it making no need of irq for SPI. GPIO pins and manager for sure are needed, I already started to look at them, as well as IRQ support (not for SPI but to handle the incoming IRQ line of the CAN device).
If the concept of having a generic HAL on one side and Wrapper to Circle that would be the 'visible part' of the lib on Circle side is ok for you then I will go on that way.  For the CAN HAL itself I believe I don't need much support, as I m used since long time to program such kind of stuff on different MCU, but for the Circle side for the wrapper most probably I will need some help at some point.

For the moment I'm working on UML model file (.xmi  file in my repo) to structure the whole thing based on the concept I exposed in the 3 first specifications files (sorry for my _old school_ way of doing programming, usually spending long time on paper or on such modelling systems before starting to write fisr line of code). 
->I think I will need about up to 1 month to complete this preparation phase. Once it is ready the code will be generated automatically form this UML model for all structures, objects and functions prototypes. I created this .xmi file with Umbrello (available on kde.org) but I will put also the export as html page on the repo to allow looking at it without installing Umbrello.

For info I activated also discussions on my repo.

**Reply by rsta2** [reply]
Understood. A quick note on IRQ generation: Because the IRQ line of the CAN device is connected via a GPIO pin, it has to be handled by the class `CGPIOPin` in Circle. This includes the IRQ handling, which requires also to use the class `CGPIOManager`. Normally you do not need to care about IRQ vectors etc. here. This is all handled by these classes for GPIO interrupts.

**Reply by rsta2** [reply]
I have installed Umbrello here today, so I can view the .xmi file, without the need of converting it to .html. I did not use UML that far for my projects and not for generating code, but if you prefer this, it's OK of course. I think I can read your models, where it is necessary. I'm fine with, what I have seen so far.

**Reply by NSD76** [reply]
Thanks for info, currently there are only 2 classes and 1 interface (the HAL), I will progress on this in the coming weeks.
I Will anyway post again some notice once I have a much more advanced status that could be reviewed.

**Comment by rsta2** [comment]
Version 1.0.0 of the [CAN library for Circle](https://github.com/NSD76/CANlib-MCP2515-for-Circle) by @NSD76 has been released.

Please see the [Announcement](https://github.com/NSD76/CANlib-MCP2515-for-Circle/discussions/6)!
